src/com/ms/silverking/cloud/dht/net/IncomingMessageGroup.java

```diff
@@ -236,182 +243,182 @@ public ReadResult readFromChannel(SocketChannel channel) throws IOException {
+          } else {
+            lastNumRead += numRead;
+            if (curBuffer.remaining() == 0) {
+              curBufferIndex++;
+              assert curBufferIndex <= buffers.length;
+              if (curBufferIndex == buffers.length) {
+                readState = ReadState.DONE;
+                return ReadResult.COMPLETE;
+              } else {
+                break;
+              }
+            } else {
+              break;
             }
+          }
+        case DONE:
+          if (debug) {
+            Log.info("IncomingBufferedData.DONE");
+          }
+          return ReadResult.COMPLETE;
+        case CHANNEL_CLOSED:
+          throw new IOException("Channel closed");
+        default:
+          throw new RuntimeException("panic");
         }
-    }
-    
-    public String toString() {
-        StringBuilder  sb;
-        
-        sb = new StringBuilder();
-        sb.append("*********************************\n");
-        sb.append(bufferToString(bufferLengthsBuffer));
-        sb.append("buffers.length\t"+ buffers.length);
-        for (ByteBuffer buffer : buffers) {
-            sb.append(bufferToString(buffer));
+      } catch (IOException ioe) {
+        if (debug) {
+          Log.logErrorWarning(ioe);
         }
-        sb.append(lastNumRead +"\t");
-        sb.append(lastNumRead);
-        sb.append(readState);
-        sb.append("\n*********************************\n");
-        return sb.toString();
-    }
-    
-    public String bufferToString(ByteBuffer buf) {
-        if (buf == null) {
-            return "[null]";
+        if (ioe.getMessage().startsWith("Connection reset")) {
+          readState = ReadState.CHANNEL_CLOSED;
+          return ReadResult.CHANNEL_CLOSED;
         } else {
-            StringBuilder  sb;
-            ByteBuffer     dup;
-            
-            dup = buf.duplicate();
-            dup.rewind();
-            sb = new StringBuilder();
-            sb.append('[');
-            sb.append(dup.limit());
-            sb.append('\t');
-            while (dup.remaining() > 0) {
-                byte   b;
-                
-                b = dup.get();
-                sb.append(Integer.toHexString(b) +":");
-            }
-            sb.append(']');
-            return sb.toString();
+          throw ioe;
         }
+      }
+    } while (true);
+  }
+
+  private void allocateBufferLengthsBuffer(int numBuffers) throws IOException {
+    if (debug) {
+      Log.fine("allocateBufferLengthsBuffer ", numBuffers);
+    }
+    if (numBuffers < minNumBuffers) {
+      throw new IOException("numBuffers < " + minNumBuffers);
+    }
+    if (numBuffers > maxNumBuffers) {
+      throw new IOException("numBuffers > maxNumBuffers\t" + numBuffers + " > " + maxNumBuffers);
+    }
+    try {
+      bufferLengthsBuffer = ByteBuffer.allocate(numBuffers * NumConversion.BYTES_PER_INT);
+    } catch (OutOfMemoryError oome) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)